text/0000-no-entry.md

```diff
@@ -0,0 +1,149 @@
+in sandbox environments.
+
+The target toolchain will commonly supply an entry-point symbol (e.g.
+`_start`) by default. As a result, attempting to define an entry-point symbol
+typically results in a link error. Addressing this currently requires linking
+manually with an toolchain-specific argument like `-nostartfiles`.
+
+This proposal introduces a top-level attribute `no_entry`, which will cause the
+Rust toolchain to omit the entry-point symbol. Specifying `#![no_entry]` allows
+(and requires) the user to supply such a symbol.
+
+(Please note that this is entirely distinct from the Rust-supplied startup hook
+`#[start]`.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Specifying the attribute `#![no_entry]` at the top level of a Rust binary crate
+will cause Rust to omit the entry-point symbol when compiling and linking the
+binary. When linking, Rust will pass any necessary target-specific arguments to
+the toolchain to omit the target's entry-point code, such as `-nostartfiles`.
+
+Binaries built with this attribute set will typically not link unless the user
+supplies their own entry-point symbol.
+
+Libraries cannot use this attribute; attempting to do so will produce a
+compile-time error stating that `no_entry` only applies to binaries.
+
+A binary built with this attribute set could supply the entry-point symbol
+itself, or have the entry-point symbol supplied by a library crate or by a
+non-Rust library.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Until stabilized, the `no_entry` attribute will require a feature-gate named
+`no_entry`.
+
+The `no_entry` attribute may be specified as part of a `cfg_attr` declaration,
+to omit the entry point in some configurations but not others.
+
+Specifying the `no_entry` attribute on anything other than the top-level module
+of a binary crate will produce an error (`warning: crate-level attribute should
+be in the root module`).
+
+Declaring an entry-point symbol will typically require a declaration with
+`#[no_mangle] pub extern "C"`.
+
+To implement `no_entry`, the compiler can have a default linker argument based
+on the type of linker in use, such as `-nostartfiles` for GCC or clang/LLD.
+Specific targets which require more specific options can provide those options
+via the target configuration structures.
+
+Some targets may be able to support invoking the C library without running the
+startup code provided by the C library; other targets may not support this, or
+have limitations on such support. Targets that do not support this should
+document the target-specific limitations of `no_entry` for their target, may
+require `no_std` in addition to `no_entry`, or may not support `no_entry` at
+all.
+
+On targets that already don't provide an entry point, the compiler will
+silently accept `no_entry`. On targets that cannot support `no_entry`, the
+compiler will emit a compile-time error.
+
```

Comment:
Nitpick: I’d prefer if both cases were an error.  Options that have no effect are confusing and tend to lead to cargo-cult programming.

Reply:
I disagree -- I believe that if the target provides no entry and you have `#![no_entry]` with no effect then the compiler has fulfilled its contract. I think it makes sense from a declarative specification of behavior (and `#![no_entry]` is a description, not a command).

Reply:
What @Centril said. I'd like to avoid having to special-case platforms that never have an entry point, when writing code that should never have an entry point. You wrote `no_entry` and the program will have no entry, so the compiler can just say "OK, done, that was easy".

Reply:
@joshtriplett It would be good to reflect resolving the conversation in the text also with the discussion as material. :)